Lib/asyncio/taskgroups.py

```diff
@@ -54,21 +54,22 @@ async def __aenter__(self):
 
     async def __aexit__(self, et, exc, tb):
         self._exiting = True
-        propagate_cancellation_error = None
 
         if (exc is not None and
                 self._is_base_error(exc) and
                 self._base_error is None):
             self._base_error = exc
 
-        if et is not None:
-            if et is exceptions.CancelledError:
-                if self._parent_cancel_requested and not self._parent_task.uncancel():
-                    # Do nothing, i.e. swallow the error.
-                    pass
-                else:
-                    propagate_cancellation_error = exc
+        propagate_cancellation_error = \
+            exc if et is exceptions.CancelledError else None
+        if self._parent_cancel_requested:
+            # If this flag is set we *must* call uncancel().
+            if self._parent_task.uncancel() == 0:
```

Comment:
there's now an issue where `uncancel() != 0` and there's no CancelledError to propagate:

```python
import asyncio
import contextlib

class CustomException(Exception):
    pass

async def raise_exc():
    raise CustomException

@contextlib.asynccontextmanager
async def database():
    try:
        yield
    finally:
        raise CustomException


async def main():
    try:
        async with asyncio.timeout(0.1):
            try:
                async with asyncio.TaskGroup():
                    async with database():
                        await asyncio.sleep(10)
            except* CustomException:
                 print("custom exception happened!")
    except TimeoutError as e:
        print("timed out correctly")
    else:
        assert False, "failed to timeout"

asyncio.run(main())
```

Reply:
It is failing even on main without this PR so is unrelated, you should probably create a new issue for this.

Reply:
this one is happens on both main and this PR

```python
import asyncio
import contextlib
import logging

logger = logging.getLogger(__name__)

class CustomerDataCorrupted(Exception):
    pass

async def raise_exc():
    raise CustomerDataCorrupted

@contextlib.asynccontextmanager
async def database():
    try:
        yield
    finally:
        print("customer error happened, but we don't have the context to deal with it here")
        raise CustomerDataCorrupted

async def with_database():
    async with database():
        await asyncio.sleep(10)


async def main():
    try:
        async with asyncio.timeout(0.1):
            try:
                async with asyncio.TaskGroup() as tg:
                    tg.create_task(with_database())
            except* CustomerDataCorrupted:
                 print("the customer lost data, we need to phone them up to fix it")
    except TimeoutError:
        pass

asyncio.run(main())
```

Reply:
You are being a bit dramatic, Thomas. Both examples contain the same bug -- they have a context manager that accidentally replaces `CancelledError` with some other exception. The fix is something like
```
        if not isinstance(sys.exc_info()[1], asyncio.CancelledError):
            raise CustomException
```

Reply:
Marking as resolved as it is unrelated, also you should post unrelated issues which fail even on main to the issue not PRs.